src/app/data-model-provider/MetadataSearch.h

```diff
@@ -0,0 +1,133 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include "app/ConcreteAttributePath.h"
+#include "app/ConcreteClusterPath.h"
+#include "lib/core/DataModelTypes.h"
+#include "lib/support/CodeUtils.h"
+#include <app/data-model-provider/MetadataList.h>
+#include <app/data-model-provider/MetadataTypes.h>
+
+#include <algorithm>
+#include <optional>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+/// Helps search for a specific server cluster in the given
+/// metadata provider.
+///
+/// Facilitates the very common operation of "find a cluster with a given ID".
+class ServerClusterFinder
+{
+public:
+    ServerClusterFinder(ProviderMetadataTree * provider) : mProvider(provider) {}
+
+    std::optional<ServerClusterEntry> Find(const ConcreteClusterPath & path)
```

Comment:
Why inline?

Reply:
First draft. Moved it to actual impl to save flash (and it worked).